File: src/main/ws.js
'use strict';

const EventEmitter = require('node:events');
const WebSocket = require('ws');

const api = require('./api');

const RECONNECT_MIN = 1000;
const RECONNECT_MAX = 15000;

/**
 * Realtime link to the API server's /api/v1/ws endpoint.
 *
 * Emits:
 *   'status'  ({ state, message })  state: connecting|connected|offline|unauthorized|denied
 *   'message' (parsed payload from the server)
 */
class RealtimeClient extends EventEmitter {
  constructor() {
    super();
    this.socket = null;
    this.timer = null;
    this.attempt = 0;
    this.stopped = false;
    this.state = 'connecting';
  }

  setStatus(state, message) {
    // Re-emit even when unchanged for terminal states, so a manual retry
    // always produces visible feedback in the renderer.
    this.state = state;
    this.emit('status', { state, message });
  }

  start() {
    this.stopped = false;
    this.connect();
  }

  /** Manual retry from the tray or the renderer's setup screen. */
  retry() {
    this.attempt = 0;
    this.stopped = false;
    this.teardown();
    clearTimeout(this.timer);
    this.connect();
  }

  stop() {
    this.stopped = true;
    clearTimeout(this.timer);
    this.teardown();
  }

  teardown() {
    if (!this.socket) return;
    const socket = this.socket;
    this.socket = null;
    socket.removeAllListeners();
    try {
      socket.terminate();
    } catch {
      /* already gone */
    }
  }

  scheduleReconnect() {
    if (this.stopped) return;
    clearTimeout(this.timer);
    const delay = Math.min(RECONNECT_MIN * 2 ** this.attempt, RECONNECT_MAX);
    this.attempt += 1;
    this.timer = setTimeout(() => this.connect(), delay);
  }

  async connect() {
    if (this.stopped) return;
    this.teardown();

    // Only advertise "connecting" on the first attempt or after a manual retry.
    // Once we know we are offline, staying offline while retrying in the
    // background beats flickering between the two on every backoff tick.
    if (this.attempt === 0 && this.state !== 'connected') this.setStatus('connecting');

    let token;
    try {
      token = await api.ensureToken();
    } catch (err) {
      if (err.code === 'DENIED') {
        // Do not re-open the authorisation dialog on a loop; wait for a manual retry.
        this.stopped = true;
        this.setStatus('denied', err.message);
        return;
      }
      this.setStatus(err.code === 'OFFLINE' ? 'offline' : 'unauthorized', err.message);
      this.scheduleReconnect();
      return;
    }

    const url = `${api.wsBase()}${api.API}/ws?token=${encodeURIComponent(token)}`;
    let socket;
    try {
      socket = new WebSocket(url, { handshakeTimeout: 5000 });
    } catch (err) {
      this.setStatus('offline', err.message);
      this.scheduleReconnect();
      return;
    }
    this.socket = socket;

    socket.on('open', () => {
      this.attempt = 0;
      this.setStatus('connected');
    });

    socket.on('message', (raw) => {
      let payload;
      try {
        payload = JSON.parse(raw.toString());
      } catch {
        return;
      }
      this.emit('message', payload);
    });

    socket.on('close', (code) => {
      if (this.socket !== socket) return;
      this.socket = null;
      if (code === 1008) {
        // Server rejected our token — drop it so the next attempt re-authorises.
        api.clearToken();
        this.setStatus('unauthorized', 'The API server rejected the saved token');
      } else {
        // A cached token makes ensureToken() skip the network entirely, so for
        // an unreachable server this close is the *only* signal we get. Report
        // offline whether or not we ever managed to connect.
        this.setStatus(
          'offline',
          this.state === 'connected' ? 'Connection closed' : 'Could not reach the API server',
        );
      }
      this.scheduleReconnect();
    });

    socket.on('error', () => {
      // 'close' always follows and handles both the status and the reconnect.
    });
  }
}

module.exports = { RealtimeClient };

File: src-tauri/src/ws.rs
//! Realtime link to the API server's `/api/v1/ws` endpoint.
//!
//! The connection state is the app's connection state: once a token is cached,
//! `ensure_token` returns without touching the network, so a dead API server
//! produces **no error there at all**. The socket's close is the only signal —
//! which is why the close handler reports `offline` regardless of whether a
//! connection was ever established. Getting this wrong left the widget stuck on
//! "Connecting…" forever.

use std::sync::atomic::{AtomicBool, AtomicU32, Ordering};
use std::sync::Arc;
use std::time::Duration;

use futures_util::StreamExt;
use tauri::{AppHandle, Manager};
use tokio::sync::Notify;
use tokio_tungstenite::tungstenite::Message;

use crate::api::ErrorCode;
use crate::state::Core;

const RECONNECT_MIN_MS: u64 = 1000;
const RECONNECT_MAX_MS: u64 = 15000;
const HANDSHAKE_TIMEOUT: Duration = Duration::from_secs(5);

pub struct Realtime {
    /// Pulsed by the tray's Reconnect, the setup screen's Retry, and by any
    /// command that discovers the server has gone away.
    wake: Notify,
    /// Set after a denied authorisation so we do not re-open the dialog on a
    /// loop; only a manual retry clears it.
    parked: AtomicBool,
    attempt: AtomicU32,
}

impl Realtime {
    pub fn new() -> Self {
        Self {
            wake: Notify::new(),
            parked: AtomicBool::new(false),
            attempt: AtomicU32::new(0),
        }
    }

    /// Manual retry: drop any backoff, unpark, and reconnect now.
    pub fn request_retry(&self) {
        self.attempt.store(0, Ordering::SeqCst);
        self.parked.store(false, Ordering::SeqCst);
        self.wake.notify_waiters();
    }

    fn backoff(&self) -> Duration {
        let attempt = self.attempt.fetch_add(1, Ordering::SeqCst);
        let delay = RECONNECT_MIN_MS
            .saturating_mul(1u64 << attempt.min(8))
            .min(RECONNECT_MAX_MS);
        Duration::from_millis(delay)
    }

    /// Sleep, unless a manual retry arrives first.
    async fn wait(&self, delay: Duration) {
        tokio::select! {
            _ = tokio::time::sleep(delay) => {}
            _ = self.wake.notified() => {}
        }
    }
}

/// Ask the realtime task to reconnect now. A no-op before setup finishes.
pub fn retry(app: &AppHandle) {
    if let Some(realtime) = app.try_state::<Arc<Realtime>>() {
        realtime.request_retry();
    }
}

pub fn spawn(core: Arc<Core>, realtime: Arc<Realtime>) {
    tauri::async_runtime::spawn(async move {
        loop {
            // Parked after a denied authorisation: wait for a manual retry.
            if realtime.parked.load(Ordering::SeqCst) {
                realtime.wake.notified().await;
                continue;
            }

            match connect_once(&core, &realtime).await {
                Outcome::Reconnect => {
                    let delay = realtime.backoff();
                    realtime.wait(delay).await;
                }
                Outcome::Parked => {}
            }
        }
    });
}

enum Outcome {
    Reconnect,
    Parked,
}

async fn connect_once(core: &Arc<Core>, realtime: &Arc<Realtime>) -> Outcome {
    // Only advertise "connecting" on the first attempt or after a manual retry.
    // Once we know we are offline, staying offline while retrying in the
    // background beats flickering between the two on every backoff tick.
    if realtime.attempt.load(Ordering::SeqCst) == 0 && core.status() != "connected" {
        core.set_status("connecting", "");
    }

    let token = match core.api.ensure_token().await {
        Ok(token) => token,
        Err(err) => {
            if err.code == ErrorCode::Denied {
                realtime.parked.store(true, Ordering::SeqCst);
                core.set_status("denied", &err.message);
                return Outcome::Parked;
            }
            let status = if err.code == ErrorCode::Offline {
                "offline"
            } else {
                "unauthorized"
            };
            core.set_status(status, &err.message);
            return Outcome::Reconnect;
        }
    };

    let url = format!(
        "{}{}/ws?token={}",
        core.store.ws_base(),
        crate::api::API,
        urlencode(&token)
    );

    let socket = match tokio::time::timeout(HANDSHAKE_TIMEOUT, tokio_tungstenite::connect_async(url))
        .await
    {
        Ok(Ok((socket, _response))) => socket,
        Ok(Err(err)) => {
            core.set_status("offline", &format!("Could not reach the API server: {err}"));
            return Outcome::Reconnect;
        }
        Err(_) => {
            core.set_status("offline", "Could not reach the API server");
            return Outcome::Reconnect;
        }
    };

    realtime.attempt.store(0, Ordering::SeqCst);
    core.set_status("connected", "");

    let (_sink, mut stream) = socket.split();
    let mut policy_rejection = false;

    loop {
        let next = tokio::select! {
            message = stream.next() => message,
            // A manual retry tears the socket down rather than waiting for it
            // to notice on its own.
            _ = realtime.wake.notified() => None,
        };

        match next {
            Some(Ok(Message::Text(raw))) => {
                if let Ok(payload) = serde_json::from_str::<serde_json::Value>(&raw) {
                    core.handle_message(&payload).await;
                }
            }
            Some(Ok(Message::Binary(raw))) => {
                if let Ok(payload) = serde_json::from_slice::<serde_json::Value>(&raw) {
                    core.handle_message(&payload).await;
                }
            }
            Some(Ok(Message::Close(frame))) => {
                policy_rejection = frame.map(|frame| u16::from(frame.code) == 1008).unwrap_or(false);
                break;
            }
            Some(Ok(_)) => {}
            // An error is always followed by the stream ending; both land here.
            Some(Err(_)) | None => break,
        }
    }

    if policy_rejection {
        // The server rejected our token — drop it so the next attempt re-authorises.
        core.api.clear_token();
        core.set_status("unauthorized", "The API server rejected the saved token");
    } else {
        core.set_status(
            "offline",
            if core.status() == "connected" {
                "Connection closed"
            } else {
                "Could not reach the API server"
            },
        );
    }

    Outcome::Reconnect
}

fn urlencode(value: &str) -> String {
    let mut out = String::with_capacity(value.len());
    for byte in value.as_bytes() {
        match byte {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' => {
                out.push(*byte as char)
            }
            other => out.push_str(&format!("%{other:02X}")),
        }
    }
    out
}
